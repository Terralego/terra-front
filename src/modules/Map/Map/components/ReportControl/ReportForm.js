import React, { useState, useCallback } from 'react';
import { Button, Label, RadioGroup, Radio, TextArea } from '@blueprintjs/core';

export const ReportForm = ({ url, coordinates, onSubmit, onCancel, translate: t }) => {
  const [comment, setComment] = useState('');
  const [reportType, setReportType] = useState('wrong_info');

  const onCommentChange = useCallback(e => setComment(e.currentTarget.value), []);
  const onReportTypeChange = useCallback(e => setReportType(e.currentTarget.value), []);

  const submitReport = useCallback(e => {
    e.preventDefault();
    onSubmit({ comment, coordinates, url, reportType });
  }, [onSubmit, comment, coordinates, url, reportType]);

  return (
    <form onSubmit={submitReport}>
      <input hidden id="report-geometry" value={coordinates} />
      <RadioGroup
        label={t('map.report_control.form.radiogroup.label')}
        selectedValue={reportType}
        onChange={onReportTypeChange}
      >
        <Radio
          label={t('map.report_control.form.radio.wrong_info.label')}
          value="wrong_info"
        />
        <Radio
          label={t('map.report_control.form.radio.missing_info.label')}
          value="missing_info"
        />
        <Radio
          label={t('map.report_control.form.radio.wrong_location.label')}
          value="wrong_location"
        />
        <Radio
          label={t('map.report_control.form.radio.doesnotexist.label')}
          value="does_not_exist"
        />
      </RadioGroup>
      <Label>
        {t('map.report_control.form.comment.label')}
        <span className="bp3-text-muted"> ({t('map.report_control.form.comment.optional')})</span>
        <TextArea value={comment} onChange={onCommentChange} />
      </Label>
      <Label>
        {t('map.report_control.form.url.label')}:
        <br />
        <a href={url}>{url}</a>
      </Label>

      <Button onClick={onCancel}>{t('map.report_control.form.cancel')}</Button>
      <Button type="submit" id="submitReport">{t('map.report_control.form.submit')}</Button>
    </form>
  );
};
export default ReportForm;
