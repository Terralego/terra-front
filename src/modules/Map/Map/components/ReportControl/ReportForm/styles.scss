.report-button-group > .bp3-form-content {
    display: flex;
    justify-content: space-evenly;
}

.report-button {
    text-transform: uppercase;
}

.report-url > .bp3-form-content {
    display: flex;
    flex-direction: column;
    flex-wrap: wrap;
}

.report-timestamp-id > .bp3-form-content {
    display: flex;
    justify-content: space-evenly;
}
