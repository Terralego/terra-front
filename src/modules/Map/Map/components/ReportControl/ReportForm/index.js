import React from 'react';
import { Button, FormGroup, InputGroup, Radio, TextArea } from '@blueprintjs/core';

import './styles.scss';

export const ReportForm = ({ url, coordinates, onSubmit, onCancel, translate: t }) => {
  const submitReport = e => {
    e.preventDefault();
    const data = {};
    const fd = new FormData(e.target);
    fd.forEach((value, key) => { data[key] = value; });
    onSubmit(data);
  };

  return (
    <form onSubmit={submitReport} onReset={onCancel}>
      <input hidden name="lng" value={coordinates.lng} />
      <input hidden name="lat" value={coordinates.lat} />
      <FormGroup
        label={t('terralego.map.report_control.form.radiogroup.label')}
      >
        <Radio
          label={t('terralego.map.report_control.form.radio.wrong_info.label')}
          value="wrong_info"
          name="reportType"
          defaultChecked
        />
        <Radio
          label={t('terralego.map.report_control.form.radio.missing_info.label')}
          value="missing_info"
          name="reportType"
        />
        <Radio
          label={t('terralego.map.report_control.form.radio.wrong_location.label')}
          value="wrong_location"
          name="reportType"
        />
        <Radio
          label={t('terralego.map.report_control.form.radio.doesnotexist.label')}
          value="does_not_exist"
          name="reportType"
        />
      </FormGroup>
      <FormGroup>
        {t('terralego.map.report_control.form.comment.label')}
        <span className="bp3-text-muted"> ({t('terralego.map.report_control.form.comment.optional')})</span>
        <TextArea large fill name="comment" />
      </FormGroup>
      <FormGroup className="report-url">
        {t('terralego.map.report_control.form.url.label')}:
        <InputGroup large readOnly name="url" defaultValue={url} />
      </FormGroup>

      <FormGroup className="report-button-group">
        <Button
          type="reset"
          className="report-button"
        >
          {t('terralego.map.report_control.form.cancel')}
        </Button>
        <Button
          type="submit"
          id="submitReport"
          className="report-button"
        >
          {t('terralego.map.report_control.form.submit')}
        </Button>
      </FormGroup>
    </form>
  );
};
export default ReportForm;
