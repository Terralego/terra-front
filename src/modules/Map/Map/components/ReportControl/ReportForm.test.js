import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

import ReportForm from './ReportForm';

it('should render correctly', () => {
  const tree = renderer.create(
    <ReportForm translate={m => m} />,
  ).toJSON();

  expect(tree).toMatchSnapshot();
});

it('should return data on submit', () => {
  const submitMock = jest.fn();
  const wrapper = mount( // shallow trigger an error
    <ReportForm
      url="test/url"
      coordinates={{ lng: 42.42, lat: 31.31 }}
      onSubmit={submitMock}
      onCancel={jest.fn()}
      translate={jest.fn()}
    />,
  );

  wrapper.simulate('submit');
  expect(submitMock).toHaveBeenCalledWith({
    coordinates: { lng: 42.42, lat: 31.31 },
    url: 'test/url',
    comment: '',
    reportType: 'wrong_info', // default value set in component
  });
});
