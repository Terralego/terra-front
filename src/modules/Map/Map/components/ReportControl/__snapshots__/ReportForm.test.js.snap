// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    hidden={true}
    id="report-geometry"
  />
  <div>
    <label
      className="bp3-label"
    >
      map.report_control.form.radiogroup.label
    </label>
    <label
      className="bp3-control bp3-radio"
    >
      <input
        checked={true}
        name="Blueprint3.RadioGroup-0"
        onChange={[Function]}
        type="radio"
        value="wrong_info"
      />
      <span
        className="bp3-control-indicator"
      />
      map.report_control.form.radio.wrong_info.label
    </label>
    <label
      className="bp3-control bp3-radio"
    >
      <input
        checked={false}
        name="Blueprint3.RadioGroup-0"
        onChange={[Function]}
        type="radio"
        value="missing_info"
      />
      <span
        className="bp3-control-indicator"
      />
      map.report_control.form.radio.missing_info.label
    </label>
    <label
      className="bp3-control bp3-radio"
    >
      <input
        checked={false}
        name="Blueprint3.RadioGroup-0"
        onChange={[Function]}
        type="radio"
        value="wrong_location"
      />
      <span
        className="bp3-control-indicator"
      />
      map.report_control.form.radio.wrong_location.label
    </label>
    <label
      className="bp3-control bp3-radio"
    >
      <input
        checked={false}
        name="Blueprint3.RadioGroup-0"
        onChange={[Function]}
        type="radio"
        value="does_not_exist"
      />
      <span
        className="bp3-control-indicator"
      />
      map.report_control.form.radio.doesnotexist.label
    </label>
  </div>
  <label
    className="bp3-label"
  >
    map.report_control.form.comment.label
    <span
      className="bp3-text-muted"
    >
       (
      map.report_control.form.comment.optional
      )
    </span>
    <textarea
      className="bp3-input"
      onChange={[Function]}
      style={Object {}}
      value=""
    />
  </label>
  <label
    className="bp3-label"
  >
    map.report_control.form.url.label
    :
    <br />
    <a />
  </label>
  <button
    className="bp3-button"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    type="button"
  >
    <span
      className="bp3-button-text"
    >
      map.report_control.form.cancel
    </span>
  </button>
  <button
    className="bp3-button"
    id="submitReport"
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    type="submit"
  >
    <span
      className="bp3-button-text"
    >
      map.report_control.form.submit
    </span>
  </button>
</form>
`;
